In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv")
    #index_col="coin_id"

# Display sample data
df_market_data.head(10)

coin_id  price_change_percentage_24h  price_change_percentage_7d  \
0          bitcoin                      1.08388                     7.60278   
1         ethereum                      0.22392                    10.38134   
2           tether                     -0.21173                     0.04935   
3           ripple                     -0.37819                    -0.60926   
4     bitcoin-cash                      2.90585                    17.09717   
5      binancecoin                      2.10423                    12.85511   
6        chainlink                     -0.23935                    20.69459   
7          cardano                      0.00322                    13.99302   
8         litecoin                     -0.06341                     6.60221   
9  bitcoin-cash-sv                      0.92530                     3.29641   

   price_change_percentage_14d  price_change_percentage_30d  \
0                      6.57509                      7.67258   
1                      4.80849                      0.13169   
2                      0.00640                     -0.04237   
3                      2.24984                      0.23455   
4                     14.75334                     15.74903   
5                      6.80688                      0.05865   
6                      9.30098                    -11.21747   
7                      5.55476                     10.10553   
8                      7.28931                      1.21662   
9                     -1.86656                      2.88926   

   price_change_percentage_60d  price_change_percentage_200d  \
0                     -3.25185                      83.51840   
1                    -12.88890                     186.77418   
2                      0.28037                      -0.00542   
3                    -17.55245                      39.53888   
4                    -13.71793                      21.66042   
5                     36.33486                     155.61937   
6                    -43.69522                     403.22917   
7                    -22.84776                     264.51418   
8                    -17.23960                      27.49919   
9                    -24.87434                       7.42562   

   price_change_percentage_1y  
0                    37.51761  
1                   101.96023  
2                     0.01954  
3                   -16.60193  
4                    14.49384  
5                    69.69195  
6                   325.13186  
7                   156.09756  
8                   -12.66408  
9                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [index]   (value)

---

### Prepare the Data

In [5]:
df_market_data.dtypes

coin_id                          object
price_change_percentage_24h     float64
price_change_percentage_7d      float64
price_change_percentage_14d     float64
price_change_percentage_30d     float64
price_change_percentage_60d     float64
price_change_percentage_200d    float64
price_change_percentage_1y      float64
dtype: object

In [6]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d","price_change_percentage_30d",
                                                              "price_change_percentage_60d","price_change_percentage_200d","price_change_percentage_1y"]])



In [7]:
# Create a DataFrame with the scaled data
df_crypto= pd.DataFrame(crypto_scaled, columns= ["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d","price_change_percentage_30d",
                                                              "price_change_percentage_60d","price_change_percentage_200d","price_change_percentage_1y"])


# Copy the crypto names from the original DataFrame
df_crypto["coin_id"] = df_market_data["coin_id"] 

# Set the coin_id column as index
df_crypto.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
df_crypto.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [8]:
import os
os.environ["OMP_NUM_THREADS"] = "1"


In [9]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))


In [10]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k

# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=1)
    model.fit(crypto_scaled)
    inertia.append(model.inertia_)



In [11]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow

k     inertia
0   1  287.000000
1   2  212.123342
2   3  165.136752
3   4   79.022435
4   5   66.413051
5   6   61.634088
6   7   46.074467
7   8   43.392694
8   9   32.913223
9  10   30.469111

In [12]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  In this plot, the elbow appears to occur around "K= 4" because the decrease in inertia becomes less significant after this point. This indicates that 4 clusters might be an optimal choice for your data.


---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [13]:
import os
os.environ["OMP_NUM_THREADS"] = "1"


In [14]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=3)


In [15]:
# Fit the K-Means model using the scaled DataFrame
model.fit(crypto_scaled)


KMeans(n_clusters=4, random_state=3)

In [16]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
from sklearn.cluster import KMeans
k_3 = model.predict(crypto_scaled)


# Print the resulting array of cluster values.
print(k_3)

[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 3 1 0 0 2
 0 0 0 0]


In [17]:
# Create a copy of the scaled DataFrame
copy_df = crypto_scaled.copy()

In [18]:
import pandas as pd

# Convert crypto_scaled to a DataFrame if it's not already one
if not isinstance(crypto_scaled, pd.DataFrame):
    copy_df = pd.DataFrame(
        crypto_scaled,
        columns=[
            "price_change_percentage_24h", 
            "price_change_percentage_7d", 
            "price_change_percentage_14d",
            "price_change_percentage_30d",
            "price_change_percentage_60d",
            "price_change_percentage_200d",
            "price_change_percentage_1y"
        ]
    )
else:
    copy_df = crypto_scaled.copy()

# Ensure k_3 matches the DataFrame dimensions
if k_3.ndim > 1:
    k_3 = k_3.flatten()

# Add the clusters column
copy_df['clusters'] = k_3

# Display the first few rows
print(copy_df.tail())


# Add a new column to the copy of the scaled DataFrame with the predicted clusters

#copy_df['clusters'] = k_3

# Display the copy of the scaled DataFrame
#copy_df.head()


    price_change_percentage_24h  price_change_percentage_7d  \
36                     1.045530                   -0.618328   
37                    -0.409044                   -0.906963   
38                     0.414711                    0.414044   
39                     0.078038                   -0.687745   
40                     1.217453                   -0.607714   

    price_change_percentage_14d  price_change_percentage_30d  \
36                     2.907054                     5.351455   
37                    -1.298986                    -1.393153   
38                    -0.047386                    -0.465380   
39                    -0.009191                    -0.058214   
40                    -0.907066                     0.449939   

    price_change_percentage_60d  price_change_percentage_200d  \
36                     4.769913                      3.148875   
37                    -0.696937                     -0.422835   
38                     0.128185          

In [19]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
import hvplot.pandas
scatter_plot = copy_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="clusters",  # Color by clusters
    hover_cols=["crypto_name"],  # Add crypto names as hover information
    title="Cryptocurrency Clustering by Price Changes",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    height=400,
    width=600,
)
scatter_plot



:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,clusters)

---

### Optimize Clusters with Principal Component Analysis.

In [20]:
# Create a PCA model instance and set `n_components=3`.
from sklearn.decomposition import PCA
pca = PCA(n_components=3)

In [21]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
crypto_pca = pca.fit_transform(crypto_scaled)

# View the scaled PCA data
crypto_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [22]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance_ratio = pca.explained_variance_ratio_
print("Explained variance ratio for each component:", explained_variance_ratio)


Explained variance ratio for each component: [0.3719856  0.34700813 0.17603793]


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.371+0.347+0.176 = 0.89 So, the total explained variance of the three principal components would be 0.89 or 89% of the total variance in the dataset.



In [23]:

#coin_ids = df_crypto['coin_id'].values  # Extract coin_ids as an array

# Create a new DataFrame with the PCA data
crypto_pca_df = pd.DataFrame(
    crypto_pca,
    columns=["PCA1", "PCA2", "PCA3"]
)

# Copy the coin_ids from the index of df_crypto
crypto_pca_df['coin_id'] = df_crypto.index.copy()

# Set the coin_id column as index
crypto_pca_df.set_index('coin_id', inplace=True)

# Display the scaled PCA DataFrame
print(crypto_pca_df.head())



                  PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715


---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [24]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))


In [25]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=1)
    model.fit(crypto_pca_df)
    inertia.append(model.inertia_)



In [26]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow

k     inertia
0   1  256.874086
1   2  182.339530
2   3  135.442408
3   4   49.665497
4   5   38.672582
5   6   34.507577
6   7   23.728365
7   8   18.574523
8   9   15.572672
9  10   11.407667

In [27]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k when using the PCA data is 4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it's the same value for K.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [28]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)


In [29]:
# Fit the K-Means model using the PCA data
model.fit(crypto_pca_df)

KMeans(n_clusters=4, random_state=1)

In [30]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
k = model.predict(crypto_pca_df)

# Print the resulting array of cluster values.
print(k)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [31]:
# Create a copy of the scaled PCA DataFrame
pca_copy_df = crypto_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
pca_copy_df["clusters"]= k

# Display the copy of the scaled PCA DataFrame
pca_copy_df.head()

PCA1      PCA2      PCA3  clusters
coin_id                                             
bitcoin      -0.600667  0.842760  0.461595         2
ethereum     -0.458261  0.458466  0.952877         2
tether       -0.433070 -0.168126 -0.641752         0
ripple       -0.471835 -0.222660 -0.479053         0
bitcoin-cash -1.157800  2.041209  1.859715         2

In [32]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_copy_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="clusters"
)

:NdOverlay   [clusters]
   :Scatter   [PCA1]   (PCA2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [39]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!
original_elbow_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Original Elbow Curve", 
    xticks= list(df_elbow["k"].unique()),
    width=500, 
    height=400
)

# Generate the PCA Elbow curve
pca_elbow_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="PCA Elbow Curve", 
    xticks=list(df_elbow["k"].unique()),
    width=500, 
    height=400
)

# Display both plots for comparison
original_elbow_plot + pca_elbow_plot





:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [40]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
original_cluster_plot = copy_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="clusters",
    hover_cols=["coin_id"],
    title="Original Clusters",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    width = 500,
    height = 400
)

# Generate the scatter plot for the PCA data
pca_cluster_plot = pca_copy_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    c="clusters",
    hover_cols=["coin_id"],
    title="PCA Clusters",
    xlabel="Principal Component 1",
    ylabel="Principal Component 2",
    width =500,
    height = 400
)

# Display both plots for comparison
original_cluster_plot + pca_cluster_plot




:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,clusters)
   .Scatter.II :Scatter   [PCA1]   (PCA2,clusters,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  Using fewer features by applying PCA reduces the dimensionality, which simplifies the clustering process and improves computational efficiency. The clusters may remain similar to those generated using the original data, but PCA ensures that the most critical variance is captured. This often leads to easier visualization and interpretation without significant loss of information.






